android/KMEA/app/src/main/java/com/tavultesoft/kmea/LanguageListActivity.java

```diff
@@ -414,11 +420,53 @@ protected void onPostExecute(JSONObject jsonObj) {
       }
 
       try {
-        languages = jsonObj.getJSONObject(KMKeyboardDownloaderActivity.KMKey_Languages).getJSONArray(KMKeyboardDownloaderActivity.KMKey_Languages);
-        options = jsonObj.getJSONObject(KMKeyboardDownloaderActivity.KMKey_Options);
         keyboardsInfo = new HashMap<String, HashMap<String, String>>();
         keyboardModifiedDates = new HashMap<String, String>();
 
+        if (!hasConnection) {
+          // When offline, only use keyboards available from kmp.json
+          // Use default options
+          options = JSONUtils.defaultOptions(context.getString(R.string.device_type));
+          languages = kmpLanguagesArray;
+        } else {
+          // Otherwise, merge kmpLanguagesArray with cloud languagesArray
+          options = jsonObj.getJSONObject(KMKeyboardDownloaderActivity.KMKey_Options);
+          languages = jsonObj.getJSONObject(KMKeyboardDownloaderActivity.KMKey_Languages).getJSONArray(KMKeyboardDownloaderActivity.KMKey_Languages);
+          for (int i = 0; i < kmpLanguagesArray.length(); i++) {
+            JSONObject kmpLanguage = kmpLanguagesArray.getJSONObject(i);
+            String kmpLanguageID = kmpLanguage.getString("id");
+            int languageIndex = JSONUtils.findID(languages, kmpLanguageID);
+            if (languageIndex == -1) {
+              // Add new language object
+              languages.put(kmpLanguage);
+            } else {
+              // Merge language info
+              JSONObject language = languages.getJSONObject(languageIndex);
+              JSONArray keyboards = language.getJSONArray("keyboards");
+              JSONArray kmpKeyboards = kmpLanguage.getJSONArray("keyboards");
+              for (int j = 0; j < kmpKeyboards.length(); j++) {
+                JSONObject kmpKeyboard = kmpKeyboards.getJSONObject(j);
+                String kmpKeyboardID = kmpKeyboard.getString("id");
+                int keyboardIndex = JSONUtils.findID(keyboards, kmpKeyboardID);
+                if (keyboardIndex == -1) {
+                  // Add new keyboard object
+                  keyboards.put(kmpKeyboard);
+                } else {
+                  // Merge keyboard info
+                  JSONObject keyboard = keyboards.getJSONObject(keyboardIndex);
+
+                  String keyboardVersion = keyboard.getString(KMManager.KMKey_KeyboardVersion);
+                  String kmpKeyboardVersion = kmpKeyboard.getString(KMManager.KMKey_KeyboardVersion);
+                  if (FileUtils.compareVersions(kmpKeyboardVersion, keyboardVersion) == FileUtils.VERSION_GREATER) {
```

Comment:
I think greater than or equal to (no need to download from online if we have same version offline)?